Fix issue #108 "OS X: java process keeps running after closing the last Logisim-evolution window"

maehne opened the pull request:
This pull request should fix issue #108 "OS X: java process keeps running after closing the last Logisim-evolution window". I builds on PR #213 and consists of the rebased patches by @davidhutchens. In consequence, at least JDK 9 is now needed to build Logisim-evolution.

@davidhutchens: Could you please check whether I completely applied your modifications and whether it works for you?

BFH-ktt1 commented:
Sorry, I mistook this pull request for the one on master, my mistake

davidhutchens commented:
Is the above "All checks have failed" message because you are compiling tests with Java 8?

I look at the files shows the changes I expect. However, MRJAdapter.jar seems to have reappeared in the lib directory. Moving it out does not cause me any problems.

I built a Mac app from this version using Java 11.0.2. Thanks @maehne for the app build capability!

It works as I expected. It fixes the specific issue #108. It puts Logisim-evolution in the menu bar as the app name. It also allows me to open files by double-clicking on them in the Finder.

When I build a jar file with "gradle jar", the resulting jar fails to run from the command line with "java -jar logisim-evolution-2.15.0.jar" giving the message "no main manifest attribute, in logisim-evolution-2.15.0.jar". Curiously, when I go into the app at "Logisim-evolution.app/Contents/Java" and run the jar found there with the same command, it works!

I notice that this version runs a simulation I have at about 1/2 the speed of version 2.14.2 (with these same changes made). I strongly suspect some other changes between the two, but I haven't checked to see if I can determine which version slowed down.

BFH-ktt1 commented:
@davidhutchens :
Concerning the gradle issue, I observed the same thing. Using "gradle shadowjar" you get a jar file that is called "logisim-evolution-2.15.0-all.jar" that executes. As I am quite new to gradle I have to look into details.
Concerning the simulation speed issue, I might have a clew why, as I introduced the new pin-shapes, and to be compatible with the old ones, I had to introduce very dirty hacks to prevent some nasty side-effects when switching between the two shape forms (e.g. each time a pin or probe is drawn, it's bounding box is newly determined). To see if the problem is due to this part you can mark out in com.cburch.logisim.std.wiring:
pin.java line 782:  painter.getInstance().recomputeBounds();
probe.java line 385: painter.getInstance().recomputeBounds();
If this solves the problem I would suggest to remove those two lines and to make a remark in the setting that when changing the pin-shapes logisim needs to be restarted to prevent incorrect redrawing of the symbols after changing this setting.

maehne commented:
@davidhutchens: Thanks for your review of the PR! I have removed the `MRJAdapter.jar` and adjusted the Travis CI configuration to now build using `openjdk9`, `oraclejdk9`, `openjdk11`, and `oraclejdk11`. As @BFH-ktt1 pointed out: only `gradle shadowJar` builds a self-contained JAR archive. The one built with `grade jar` needs the JARs of all 3rd-party dependencies in the CLASSPATH. This is desired behaviour for packaging Logisim-evolution, e.g., for Linux distributions.

@BFH-ktt1 and @alexandremalki: What do you think about building just against `openjdk9` and `openjdk11` and to drop the `oraclejdk`s? Except for the license, they should be identical with respect to the parts that we use in Logisim-evolution. If you agree, I would push the change to `.travis.yml` in this PR. If you agree also with the other modifications in this PR, I think it is ready to get merged into `develop`.

maehne commented:
@BFH-ktt1 and @alexandremalki: I added build instructions to `README.md` and took the liberty to improve formatting of the file a bit. I also gave Travis CI a bit more thought and came to the conclusion that we should test Logisim-evolution against OpenJDK 8, 9, and 10 and updated `.travis.yml` accordingly.

davidhutchens commented:
I tried @BFH-ktt1 suggestion of commenting out those two lines in Pin.java and Probe.java. That had no noticeable change in the speed of the simulation. I tried the develop branch. It also runs slower than 2.14.2 did. I'll see if I can narrow down the place it was introduced and file an issue later.  I conclude that this PR is not related to the slowness.

I think this PR is ready to be merged.

maehne commented:
Thanks @davidhutchens for your feedback and your willingness to look further into observed simulation speed regression!

@BFH-ktt1: Could you please merge this PR?
